File: iOS/DbyPaas_Demo/DbyPaas_Demo_OC/LivePlay/LivePlayEntranceViewController.m
//
//  LivePlayEntranceViewController.m
//  DbyPaas_Demo_OC
//
//  Created by 姚晓丙 on 2019/11/6.
//  Copyright © 2019 yxibng. All rights reserved.
//

#import "LivePlayEntranceViewController.h"
#import "LivePlayViewController.h"
#import <DbyPaas_iOS/DbyPaas_iOS.h>


static NSString *kUid = @"dby_uid";
static NSString *kChannelId = @"dby_channelId";

@interface LivePlayEntranceViewController ()

@property (weak, nonatomic) IBOutlet UITextField *channelIdTextField;
@property (weak, nonatomic) IBOutlet UITextField *uidTextField;

@end

@implementation LivePlayEntranceViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    NSUserDefaults *defauts = [NSUserDefaults standardUserDefaults];
    self.channelIdTextField.text = [defauts objectForKey:kChannelId];
    self.uidTextField.text = [defauts objectForKey:kUid];
    
    
}



- (IBAction)enter:(id)sender {
    
    NSString *channelId = self.channelIdTextField.text;
    NSString *uid = self.uidTextField.text;
    
    if ([self isValidUid:uid] && [self isValidChannelId:channelId]) {
        
        NSUserDefaults *defauts = [NSUserDefaults standardUserDefaults];
        [defauts setObject:channelId forKey:kChannelId];
        [defauts setObject:uid forKey:kUid];
        [defauts synchronize];
        
        
        UIStoryboard * sb = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
        LivePlayViewController *vc = [sb instantiateViewControllerWithIdentifier:@"LivePlayViewController"];
        vc.channelId = channelId;
        vc.uid = uid;
        [self.navigationController pushViewController:vc animated:YES];

    } else {
        
        //输入不合法
        
        
    }
    
    
}


- (BOOL)isValidChannelId:(NSString *)channelId
{
    if (!channelId) {
        return NO;
    }

    NSError *error;
    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:DbyChannelIdRegex options:NSRegularExpressionCaseInsensitive error:&error];
    if (error) {
        return NO;
    }
    NSArray<NSTextCheckingResult *> *results = [regex matchesInString:channelId options:NSMatchingReportCompletion range:NSMakeRange(0, channelId.length)];
    if (results.count > 0) {
        return YES;
    }
    return NO;
}

- (BOOL)isValidUid:(NSString *)uid
{
    if (!uid) {
        return NO;
    }

    NSError *error;
    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:DbyUidRegex options:NSRegularExpressionCaseInsensitive error:&error];
    if (error) {
        return NO;
    }
    NSArray<NSTextCheckingResult *> *results = [regex matchesInString:uid options:NSMatchingReportCompletion range:NSMakeRange(0, uid.length)];
    if (results.count > 0) {
        return YES;
    }
    return NO;
}
/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

@end

File: iOS/DbyPaas_Demo/DbyPaas_Demo/LivePlay/LivePlayEntranceViewController.swift
//
//  ViewController.swift
//  DbyPaas_Demo
//
//  Created by yxibng on 2019/10/29.
//  Copyright © 2019 yxibng. All rights reserved.
//

import UIKit
import DbyPaas_iOS


class LivePlayEntranceViewController: UIViewController {

    @IBOutlet weak var channelIdTextField: UITextField!
    @IBOutlet weak var uidTextField: UITextField!
    
    let channelKey = "dby_channel_key"
    let uidKey = "dby_uid_key"
    
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.channelIdTextField.text = UserDefaults.standard.object(forKey: channelKey) as? String
        self.uidTextField.text = UserDefaults.standard.object(forKey: uidKey) as? String
        
        // Do any additional setup after loading the view.
    }
    
    @IBAction func onClickEnter(_ sender: Any) {
        
        guard let channelId = channelIdTextField.text, let uid = uidTextField.text else {
            return;
        }
        
        if isValidChannelId(channelId) && isValidUid(uid) {
            let storyboard = UIStoryboard.init(name: "Main", bundle: nil)
            let videoChatVC = storyboard.instantiateViewController(withIdentifier: "DbyVideoChatViewController")
            if let chatVC = videoChatVC as? DbyVideoChatViewController {
                chatVC.channelId = channelId
                chatVC.uid = uid
                chatVC.modalPresentationStyle = .fullScreen
                present(chatVC, animated: true, completion: nil)
            }
            
            UserDefaults.standard.set(self.channelIdTextField.text, forKey: channelKey)
            UserDefaults.standard.set(uid, forKey: uidKey)
            UserDefaults.standard.synchronize()
            
        } else {
            //invalid channelId or uid
            
    
        }
    }
    
}


extension LivePlayEntranceViewController {
    func isValidChannelId(_ channelId: String) -> Bool {
        if let regex = try? NSRegularExpression.init(pattern: "^[\\w!#$%&()+-:;<=.>?@\\[\\]^_{}|~, ]{1,64}$", options: .caseInsensitive) {
            
            let results = regex.matches(in: channelId, options: .reportCompletion, range: NSRange.init(location: 0, length: channelId.count))
            
            if results.count > 0 {
                return true
            }
        }
        return false
    }
    
    func isValidUid(_ uid: String) -> Bool {
        if let regex = try? NSRegularExpression.init(pattern:"[0-9]{1,32}$" , options: .caseInsensitive) {
            
            let results = regex.matches(in: uid, options: .reportCompletion, range: NSRange.init(location: 0, length: uid.count))
            
            if results.count > 0 {
                return true
            }
        }
        return false
    }
    
}

